src/vm/dllimport.cpp

```diff
@@ -7058,29 +7081,40 @@ HINSTANCE NDirect::LoadLibraryModule( NDirectMethodDesc * pMD, LoadLibErrorTrack
             Assembly* pAssembly = pMD->GetMethodTable()->GetAssembly();
 
             SString path = pAssembly->GetManifestFile()->GetPath();
-            SString::Iterator i = path.End();
 
-            if (PEAssembly::FindLastPathSeparator(path, i))
+            SString::Iterator lastPathSeparatorIter = path.End();
+            if (PEAssembly::FindLastPathSeparator(path, lastPathSeparatorIter))
             {
-                i++;
-                path.Truncate(i);
+                lastPathSeparatorIter++;
+                path.Truncate(lastPathSeparatorIter);
 
-                path.Append(wszLibName);
+#ifdef FEATURE_PAL
```

Comment:
My suggestion would be to have a helper function that does the Load attempt (for both PAL and non-PAL) cases. For example:

hmod  = AttemptLoadLibrary(path, libname, flags, pErrorTracker);

AttemptLoadLibrary would do the loop/concatenation for actual path to check. This will simplify having the same code copied thrice (and maintaining it).